Add os_disk_encryption_set to azure_rm_virtualmachine

ephracis opened the pull request:
##### SUMMARY
Add the parameter `os_disk_encryption_set` to the
`azure_rm_virtualmachine` module, making it possible to specify which DES to use when encrypting the OS disk for a newly created VM.

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Feature Pull Request

##### COMPONENT NAME
azure_rm_virtualmachine

##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->
This is required when deploying a VM into one landing zone, using an encrypted Compute Gallery Image from another landing zone. If this is not set it will try to use the DES of the image which will fail since it belongs to another LZ.

ephracis committed:
Add os_disk_encryption_set to azure_rm_virtualmachine

Add the parameter `os_disk_encryption_set` to the
`azure_rm_virtualmachine` module, making it possible to specify which
DES to use when encrypting the OS disk for a newly created VM.

ephracis committed:
Improved checking update of DES for OSDisk

To ensure that we won't try to *add* or *change* DES on an existing VM,
the logic to check existing DES setting and comparing it to the desired
state is fixed to handle both cases.

The code is also polished to make it more readable.

Fred-sun commented on plugins/modules/azure_rm_virtualmachine.py:
```diff
@@ -206,6 +206,9 @@
             - Size of OS disk in GB.
+    os_disk_encryption_set:
+        description:
+            - ID of disk encryption set for OS disk.
```
Missing type
```suggestion
            - ID of disk encryption set for OS disk.
        type: str
```

ephracis replied:
Done

ephracis committed:
Add integration tests

Add a new kind of virtual machine tests
which creates VMs with encrypted OS disk
using the new property `os_disk_encryption_set`.

I also added the generated inventory to
`.gitignore` to avoid comitting it by mistake.

Fred-sun commented on tests/integration/targets/azure_rm_virtualmachine/tasks/setup_des.yml:
```diff
@@ -0,0 +1,58 @@
+
```
delete blank lines
```suggestion
```

ephracis replied:
Done

Fred-sun commented on tests/integration/targets/azure_rm_virtualmachine/tasks/setup_des.yml:
```diff
@@ -0,0 +1,58 @@
+    source_vault: "myvault{{ des_name }}"
+    key_url: "{{ key_url }}"
+    state: present
+  register: des_results
```
No new line character at the end of file
```suggestion
  register: des_results
```

ephracis replied:
Done

Fred-sun commented on plugins/modules/azure_rm_virtualmachine.py:
```diff
@@ -1559,6 +1565,11 @@ def exec_module(self, **kwargs):
                             vm_dict['os_profile']['linux_configuration']['disable_password_authentication']:
                         self.fail("(PropertyChangeNotAllowed) Changing property 'linuxConfiguration.disablePasswordAuthentication' is not allowed.")
 
+                current_os_des_id = vm_dict['storage_profile'].get('os_disk', {}).get('managed_disk', {}).get('disk_encryption_set', {}).get('id', {})
```
```suggestion
                current_os_des_id = vm_dict['storage_profile'].get('os_disk', {}).get('managed_disk', {}).get('disk_encryption_set', {}).get('id', None)
```

ephracis replied:
Done

Fred-sun commented on plugins/modules/azure_rm_virtualmachine.py:
```diff
@@ -1559,6 +1565,11 @@ def exec_module(self, **kwargs):
                         self.fail("(PropertyChangeNotAllowed) Changing property 'linuxConfiguration.disablePasswordAuthentication' is not allowed.")
 
+                current_os_des_id = vm_dict['storage_profile'].get('os_disk', {}).get('managed_disk', {}).get('disk_encryption_set', {}).get('id', {})
+                if self.os_disk_encryption_set is not None and current_os_des_id is not None:
```
```suggestion
                if self.os_disk_encryption_set is not None and current_os_des_id is not None and self.os_disk_encryption_set != current_os_des_id:
```

Fred-sun commented on plugins/modules/azure_rm_virtualmachine.py:
```diff
@@ -1172,6 +1177,7 @@ def __init__(self):
         self.os_type = None
         self.os_disk_caching = None
         self.os_disk_size_gb = None
+        self.os_disk_encryption_set = None
```
Please add required_if = [('os_disk_encryption_set', '*', ['managed_disk_type'])] and add "required_if=required_if" to line 1219

ephracis replied:
Done

Fred-sun commented:
> I added some integration tests to use the new disk encryption which works fine.

@ephracis  Yes, I see. But os_disk only works if managed_disk_type is not None（Please check lines 1704 through 1711）, and managed_disk_type is not specified in your use case, so it works fine. Thank you!

ephracis committed:
Polish the code

ephracis commented:
> > I added some integration tests to use the new disk encryption which works fine.
> 
> @ephracis Yes, I see. But os_disk only works if managed_disk_type is not None（Please check lines 1704 through 1711）, and managed_disk_type is not specified in your use case, so it works fine. Thank you!

Good catch! I'm adding `managed_disk_type: Standard_LRS` to that specific test case.

ephracis committed:
Fix os disk encryption in tests

Ensure that the OS disk is encrypted in the integration tests.

This requires the DES to be granted access to the KeyVault, and
that the KeyVault is protected against purges. To avoid that we
get conflicts with previously deleted, but unpurged KVs, they are
named with a date-based suffix.

ephracis commented:
Ok, I think I've managed to get the integration tests running. The problem was that I had to add the DES to the access list.

Then I ran into a second issue where I had to enable purge protection in order to use the KV for disk encryption. This means that all KVs will reside deleted but unpurged for 90 days. So in order to not get naming conflicts with these keys, they are given a date-based suffix.

Let me know if there's anything else that needs fixing. 😅

xuzhang3 commented:
@ephracis  LGTM
